tests/wsgi_test.py

```diff
@@ -1747,21 +1758,127 @@ def test_log_disable(self):
+        self.assertEqual(body, result.body)
+        self.assertEqual('/foo-bar', result.headers_lower['x-path'])
+
         self.killer.kill(KeyboardInterrupt)
         try:
             with eventlet.Timeout(1):
                 pool.waitall()
         except Exception:
             assert False, self.logfile.getvalue()
 
+    def test_close_idle_connections_listen_socket_closed(self):
+        self.reset_timeout(4)
+        self.site.application = echo_server
+        pool = eventlet.GreenPool()
+        listen_sock = eventlet.listen(('localhost', 0))
+        self.spawn_server(custom_pool=pool, sock=listen_sock,
+                          debug=True)
+        eventlet.sleep(0)  # need to enter server loop
+
+        # https://github.com/eventlet/eventlet/issues/188
+        # We want to NOT close NON-idle connections
+        sock = eventlet.connect(self.server_addr)
+        eventlet.sleep(0)
+
+        body = b'ABC' * 10
+        method_and_headers = (
+            'PUT /foo-bar HTTP/1.1\r\nHost: localhost\r\n'
+            'Content-Length: %d\r\n\r\n' % len(body))
+        sock.sendall(method_and_headers.encode('utf8'))
+        eventlet.sleep(0)
+        sock.sendall(body)
+        eventlet.sleep(0)
+        result = read_http(sock)
+        self.assertEqual('HTTP/1.1 200 OK', result.status,
+                         'Received status {0!r}; {1!r}'.format(
+                             result.status, self.logfile.getvalue()))
+        self.assertEqual(body, result.body)
+        self.assertEqual('/foo-bar', result.headers_lower['x-path'])
+        eventlet.sleep(0)
+
+        listen_sock.shutdown(socket.SHUT_RDWR)
+        eventlet.sleep(0)
+        listen_sock.close()
+        eventlet.sleep(0)
+        try:
+            with eventlet.Timeout(1):
+                pool.waitall()
+        except eventlet.Timeout:
+            eventlet.sleep(0)
+            assert False, self.logfile.getvalue()
+
+    def test_do_not_close_non_idle_connections(self):
+        self.reset_timeout(4)
+        self.site.application = echo_server
+        pool = eventlet.GreenPool()
+        listen_sock = eventlet.listen(('localhost', 0))
+        self.spawn_server(custom_pool=pool, sock=listen_sock,
+                          debug=True)
+        eventlet.sleep(0)  # need to enter server loop
+
+        # https://github.com/eventlet/eventlet/issues/188
+        # We want to NOT close NON-idle connections
+        sock = eventlet.connect(self.server_addr)
+        eventlet.sleep(0)
+
+        body = b'ABC' * 10
+        method_and_headers = (
+            'PUT /foo-bar HTTP/1.1\r\nHost: localhost\r\n'
+            'Content-Length: %d\r\n\r\n' % len(body))
+        sock.sendall(method_and_headers.encode('utf8'))
+        eventlet.sleep(0)
+
+        listen_sock.shutdown(socket.SHUT_RDWR)
+        eventlet.sleep(0)
+        listen_sock.close()
+        eventlet.sleep(0)
+        try:
+            with eventlet.Timeout(1):
```

Comment:
this is a pretty good test!  it forces the test to slow down and make sure the socket doesn't get closed.  If our test infra was running too slow it's possible we'd let a regression through, but a false negative is certainly less brittle.

Reply:
I bumped it to 2s to reduce the chance of missing a regression.